File: smoke_tests/cpp/command_injection.cpp
// Command Injection vulnerabilities in C++
#include <iostream>
#include <cstdlib>
#include <string>

void system_command(const std::string& filename) {
    // VULNERABLE: Command injection via system()
    std::string cmd = "cat " + filename;
    system(cmd.c_str());
}

void popen_command(const std::string& user_input) {
    // VULNERABLE: Command injection via popen()
    std::string command = "grep " + user_input + " /var/log/app.log";
    FILE* pipe = popen(command.c_str(), "r");
    pclose(pipe);
}

void exec_command(const char* arg) {
    // VULNERABLE: Shell execution with user input
    std::string cmd = std::string("sh -c '") + arg + "'";
    system(cmd.c_str());
}

class CommandRunner {
public:
    int run(const std::string& host) {
        // VULNERABLE: User input in shell command
        std::string cmd = "ping -c 1 " + host;
        return system(cmd.c_str());
    }
};

File: smoke_tests/php/command_injection.php
<?php
// Command Injection vulnerabilities in PHP

class CommandInjectionVulnerabilities {
    public function execCommand($userInput) {
        // VULNERABLE: Command injection via exec
        exec("ls " . $userInput);
    }

    public function systemCommand($filename) {
        // VULNERABLE: Command injection via system
        system("cat " . $filename);
    }

    public function shellExecCommand($command) {
        // VULNERABLE: Shell execution
        return shell_exec($command);
    }

    public function passthuCommand($host) {
        // VULNERABLE: Command injection via passthru
        passthru("ping -c 1 " . $host);
    }

    public function backtickCommand($arg) {
        // VULNERABLE: Backtick operator
        return `ls $arg`;
    }

    public function procOpen($cmd) {
        // VULNERABLE: proc_open with user input
        $descriptorspec = array(
            0 => array("pipe", "r"),
            1 => array("pipe", "w"),
        );
        return proc_open($cmd, $descriptorspec, $pipes);
    }

    public function popenCommand($filename) {
        // VULNERABLE: popen with user input
        return popen("cat " . $filename, "r");
    }
}
